ebu_tt_live/node/deduplicator.py

```diff
@@ -0,0 +1,199 @@
+            styles = document.binding.head.styling.style
+            document.binding.head.styling.style = None
+
+            regions = document.binding.head.layout.region
+            document.binding.head.layout.region = None
+
+            self.CollateUniqueVals(styles, _old_id_dict, _new_id_dict, hash_dict)
+            self.AppendNewElements(styles, document.binding.head.styling.style, \
+                                   _old_id_dict, _new_id_dict, hash_dict)
+
+            self.CollateUniqueVals(regions, _old_id_dict, _new_id_dict, hash_dict)
+            self.AppendNewElements(regions, document.binding.head.layout.region, \
+                                   _old_id_dict, _new_id_dict, hash_dict)
+
+            replace_id_refs = ReplaceStylesAndRegions(document.binding, \
+                                                      _old_id_dict, \
+                                                      _new_id_dict)
+            replace_id_refs.proceed()
+        else:
+            return None
+
+    def CollateUniqueVals(self, element_list, _old_id_dict, _new_id_dict, \
+                          hash_dict):
+        for value in element_list:
+            #deduplicating in-line styles
+            if value.style is not None:
+                for old_id_index in range(len(value.style)):
+                    old_id_ref = _old_id_dict.get(value.style[old_id_index])
+                    new_id_ref = _new_id_dict.get(old_id_ref)
+
+                    value.style[old_id_index] = new_id_ref
+
+            #deduplicating region elements
+            if value is not None:
+                unique_val = ComparableElement(value)
+                # stores references of original <xml:id> to <my_hash>
+                _old_id_dict[value.id] = unique_val.my_hash
+                # stores references of <my_hash> to <tt:region>
+                hash_dict[unique_val.my_hash] = value
+
+    def AppendNewElements(self, element_list, element_to_append_to, _old_id_dict, \
+                          _new_id_dict, hash_dict):
+        for hash_val in hash_dict:
+            new_id = hash_dict.get(hash_val)
+
+            for old_element in element_list:
+                if old_element.id is new_id.id:
+                    element_to_append_to.append(new_id)
+
+            _new_id_dict[hash_val] = new_id.id
+
+def ReplaceNone(none_value):
+    if none_value is None:
+        return "|" # '|' is a non-legal character and this is used to prevent
+                   # collisions between similar attributes
+    else:
+        return none_value
+
+
+class ComparableElement:
+    def __init__(self, value):
+        self.value = value
+
+        attributeDict = value._AttributeMap.copy()
+        xml_id_attr = ExpandedName('http://www.w3.org/XML/1998/namespace', 'id')
+        attributeDict.pop(xml_id_attr, None)
+
+        sortedDict = sorted(attributeDict.items(), key=lambda t: t[0])
```

Comment:
Worth a comment to explain why we're sorting it, i.e. that we want to make sure that for two elements with the same set of attributes the values are put into the hash string in the same order.